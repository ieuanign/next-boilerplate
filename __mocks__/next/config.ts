const { publicRuntimeConfig } = require("../../next.config");

const getConfig = () => ({
	publicRuntimeConfig,
	serverRuntimeConfig: publicRuntimeConfig,
});

export default getConfig;
