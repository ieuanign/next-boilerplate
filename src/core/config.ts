import getConfig from "next/config";

interface PublicRuntimeConfig {
	environment: string;
	apiHost: string;
	mockAPI: boolean;
}

const {
	publicRuntimeConfig: env,
}: { publicRuntimeConfig: PublicRuntimeConfig } = getConfig();

const apiBase = (pathname: string) => {
	return new URL(pathname, env.apiHost).toString();
};

const staticPath = (filename: string) => {
	if (filename.startsWith("/")) {
		filename = filename.substring(1);
	}

	return `/static/${filename}`;
};

export { env, apiBase, staticPath };
