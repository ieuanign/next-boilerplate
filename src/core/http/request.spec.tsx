import { rest } from "msw";
import {
	authHeaders,
	get,
	post,
	put,
	del,
	getServerSideFallbacks,
	HttpError,
} from "./request";

import { server } from "@mocks/server";
import StatusCodes from "@core/http/status-codes";
import { apiBase } from "@core/config";

describe("getServerSideFallbacks", () => {
	it("should only return the responses of successful requests", async () => {
		server.use(
			rest.get(apiBase("/api/success"), (req, res, ctx) =>
				res(ctx.status(StatusCodes.OK), ctx.json({ status: 200 }))
			)
		);
		server.use(
			rest.get(apiBase("/api/success-2"), (req, res, ctx) =>
				res(ctx.status(StatusCodes.OK), ctx.json({ status: 200 }))
			)
		);
		server.use(
			rest.get(apiBase("/api/fail"), (req, res, ctx) =>
				res(ctx.status(StatusCodes.INTERNAL_SERVER_ERROR))
			)
		);

		const res = await getServerSideFallbacks([
			[apiBase("/api/success")],
			[apiBase("/api/success-2")],
			[apiBase("/api/fail")],
		]);

		expect(res).toMatchSnapshot();
	});
});

describe("Request", () => {
	const methods: Record<string, Function> = { get, post, put, delete: del };

	describe.each(Object.keys(methods))("%s should", (key: string) => {
		const method = methods[key];
		const methodKey = key as "get" | "post" | "put" | "delete";
		it("return json response on success", async () => {
			server.use(
				rest[methodKey](apiBase("/api/success"), (req, res, ctx) =>
					res(ctx.status(StatusCodes.OK), ctx.json({ status: 200 }))
				)
			);
			const res = await method(apiBase("/api/success"));
			expect(res).toMatchSnapshot();
		});

		it("throw http error if response status is not success", async () => {
			server.use(
				rest[methodKey as "get"](apiBase("/api/fail"), (req, res, ctx) =>
					res(ctx.status(StatusCodes.BAD_REQUEST))
				)
			);

			const res = () => method(apiBase("/api/fail"));
			await expect(res).rejects.toThrow(
				new HttpError(StatusCodes.BAD_REQUEST, "Bad Request")
			);
		});
	});

	it("get with auth header should return response", async () => {
		server.use(
			rest.get(apiBase("/api/token"), (req, res, ctx) =>
				res(
					ctx.status(
						req.headers.get("Authorization")
							? StatusCodes.OK
							: StatusCodes.FORBIDDEN
					),
					ctx.json({
						status: req.headers.get("Authorization")
							? StatusCodes.OK
							: StatusCodes.FORBIDDEN,
					})
				)
			)
		);

		const res = await get(apiBase("/api/token"), {
			headers: authHeaders(),
		});

		expect(res.status).toEqual(StatusCodes.OK);
	});
});
