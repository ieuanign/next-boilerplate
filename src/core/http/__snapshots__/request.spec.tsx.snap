// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Request delete should return json response on success 1`] = `
{
  "status": 200,
}
`;

exports[`Request get should return json response on success 1`] = `
{
  "status": 200,
}
`;

exports[`Request post should return json response on success 1`] = `
{
  "status": 200,
}
`;

exports[`Request put should return json response on success 1`] = `
{
  "status": 200,
}
`;

exports[`getServerSideFallbacks should only return the responses of successful requests 1`] = `
{
  "@"http://localhost:3080/api/success",": {
    "status": 200,
  },
  "@"http://localhost:3080/api/success-2",": {
    "status": 200,
  },
}
`;
